### ISSUE TITLE
Last item in chain not referenced if preceded by Chord

### ISSUE THREAD

**Issue by rajadain** [celery/celery#3191]
## Overview

Given two sample tasks:

```python
@shared_task
def add(x, y):
    return x + y

@shared_task
def total(values):
    return sum(values)
```

And the following test case:

```python
from celery import group
from tasks import add, total

c1 = add.s(4, 4)  # Single Task
c2 = c1 | add.s(3)  # Chain of Task + Task
c3 = group(c1, c1, c1)  # Group of three Tasks
c4 = c3 | total.s()  # Chord of Group + Task
c5 = c4 | add.s(1)  # Chain of Chord + Task
c6 = c5 | add.s(1)  # Chain of Chord + Task + Task
c7 = c6 | add.s(1)  # Chain of Chord + Task + Task + Task
```

I observe the following:
```python
In [10]: c4.apply_async().get()  # Expect 24
Out[10]: 24

In [11]: c5.apply_async().get()  # Expect 25
Out[11]: 24

In [12]: c6.apply_async().get()  # Expect 26
Out[12]: 26

In [13]: c7.apply_async().get()  # Expect 27
Out[13]: 27
```

Thus, when the chord is the penultimate item in a chain with a Redis broker, calling `apply_async()` returns the Task ID of the chord, instead of the final task in the chain. This is exemplified by `c5` expecting a value of `25`, but getting `24` instead.

## Workaround

I can get around this by adding a dummy / identity task which simply returns its argument, so the chord is not the penultimate task.

## Notes

This is claimed to be fixed by #2226 but given that it is still being observed in v3.1.23, the issue remains.

Similar to #2228 and #3137. Tested with v3.1.18 and the latest v3.1.23. Observed with a Redis broker and a Django database backend.

**Comment by nicolasunravel** [comment]
using v4.0.0rc2, on python 3.4.3, I cannot reproduce this

this is the log for executing c5 from your example

```
[2016-05-11 11:23:32,694: INFO/MainProcess] Received task: tasks.add[0aaca7a1-afdc-458d-9a98-720481f52174]  
[2016-05-11 11:23:32,694: INFO/MainProcess] Received task: tasks.add[2a12338a-d6bd-44cf-ab95-65269b17d4e7]  
[2016-05-11 11:23:32,695: INFO/MainProcess] Received task: tasks.add[a59503fa-8cff-4f19-9b4d-052c0d1153c9]  
[2016-05-11 11:23:33,570: INFO/PoolWorker-2] Task tasks.add[0aaca7a1-afdc-458d-9a98-720481f52174] succeeded in 0.0030916399555280805s: 8
[2016-05-11 11:23:33,570: INFO/PoolWorker-1] Task tasks.add[a59503fa-8cff-4f19-9b4d-052c0d1153c9] succeeded in 0.003061251947656274s: 8
[2016-05-11 11:23:33,579: INFO/PoolWorker-6] Task tasks.add[2a12338a-d6bd-44cf-ab95-65269b17d4e7] succeeded in 0.012358031002804637s: 8
[2016-05-11 11:23:33,579: INFO/MainProcess] Received task: tasks.total[eda748bc-2ef2-4db4-a8e5-d1c07ce2d1af]  
[2016-05-11 11:23:33,591: INFO/MainProcess] Received task: tasks.add[db2743ea-bcd6-43d5-a13b-8368904dd79b]  
[2016-05-11 11:23:33,591: INFO/PoolWorker-7] Task tasks.total[eda748bc-2ef2-4db4-a8e5-d1c07ce2d1af] succeeded in 0.011331367073580623s: 24
[2016-05-11 11:23:33,594: INFO/PoolWorker-5] Task tasks.add[db2743ea-bcd6-43d5-a13b-8368904dd79b] succeeded in 0.0024821619736030698s: 25

```

**Comment by rajadain** [comment]
Ah I see #2226 isn't included in any main release, which is why I'm seeing this problem, which may not exist in the next version.

The log, in fact, looks the same in v3.1.23. The final task _does_ execute, but the Task ID returned by `.apply_async()` is of the penultimate task. Thus when I call `.get()`, it operates not on the final task but the penultimate one.

**Comment by dolfim** [comment]
+1
I also have the same problem with v3.1.23.

**Comment by ask** [comment]
It's fixed in 4.0rc and we have full focus on getting that released atm.
